### ISSUE TITLE
uvicorn exit with success code on error in startup stage

### ISSUE THREAD

**Issue by hongquan** [Kludex/uvicorn#780]
### Checklist

- [x] The bug is reproducible against the latest release and/or `master`.
- [x] There are no similar issues or pull requests to fix it yet.

### Describe the bug

In the stage of "startup", if some exception occurs, uvicorn just quit with exit code "0", instead of some non-zero code.

https://github.com/encode/uvicorn/blob/54d729ccc1638180a51e6ab600c4724e6424b048/uvicorn/main.py#L414-L418

With this exit code, my process manager (systemd, supervisord) cannot realize that my app exit abnormally, to restart my app.

### To reproduce

- Create a small FastAPI app with PostgreSQL database, like this example: https://fastapi.tiangolo.com/advanced/async-sql-databases/#connect-and-disconnect.

- Before running server, stop PostgreSQL, so that this code raises `ConnectionErrror`:

  ```py
  @app.on_event("startup")
  async def startup():
      await database.connect()
  ```
- Run the app with univorn.

### Expected behavior

uvicorn should quit with non-zero code.

### Actual behavior

uvicorn quit with zero code.

### Debugging material

<details>

```
$ ./runserver.py -vv
/usr/lib/python3.6/importlib/_bootstrap.py:219: RuntimeWarning: Linux supports fsync/fdsync with io_submit since 4.18 but current kernel 2.6.32
-754.el6.i686 doesn't support it. Related calls will have no effect.
  return f(*args, **kwds)
INFO:     Started server process [17004]
INFO:     Waiting for application startup.
2020-09-16 17:13:25 |INFO api.main: To connect to postgresql:///vl-agent
ERROR:    Traceback (most recent call last):
  File "/home/quan/venv/lib/python3.6/site-packages/starlette/routing.py", line 526, in lifespan
    async for item in self.lifespan_context(app):
  File "/home/quan/venv/lib/python3.6/site-packages/starlette/routing.py", line 467, in default_lifespan
    await self.startup()
  File "/home/quan/venv/lib/python3.6/site-packages/starlette/routing.py", line 502, in startup
    await handler()
  File "/home/quan/Vl-Agent/api/main.py", line 21, in startup
    await adb.connect()
  File "/home/quan/venv/lib/python3.6/site-packages/databases/core.py", line 84, in connect
    await self._backend.connect()                                                                                                       [0/175]
  File "/home/quan/venv/lib/python3.6/site-packages/databases/backends/postgres.py", line 65, in connect
    self._pool = await asyncpg.create_pool(str(self._database_url), **kwargs)
  File "/home/quan/venv/lib/python3.6/site-packages/asyncpg/pool.py", line 398, in _async__init__
    await self._initialize()
  File "/home/quan/venv/lib/python3.6/site-packages/asyncpg/pool.py", line 426, in _initialize
    await first_ch.connect()
  File "/home/quan/venv/lib/python3.6/site-packages/asyncpg/pool.py", line 125, in connect
    self._con = await self._pool._get_new_connection()
  File "/home/quan/venv/lib/python3.6/site-packages/asyncpg/pool.py", line 472, in _get_new_connection
    **self._connect_kwargs)
  File "/home/quan/venv/lib/python3.6/site-packages/asyncpg/connection.py", line 1727, in connect
    max_cacheable_statement_size=max_cacheable_statement_size)
  File "/home/quan/venv/lib/python3.6/site-packages/asyncpg/connect_utils.py", line 674, in _connect
    raise last_error
  File "/home/quan/venv/lib/python3.6/site-packages/asyncpg/connect_utils.py", line 666, in _connect
    connection_class=connection_class)
  File "/home/quan/venv/lib/python3.6/site-packages/asyncpg/connect_utils.py", line 633, in _connect_addr
    connector, timeout=timeout)
  File "/usr/lib/python3.6/asyncio/tasks.py", line 358, in wait_for
    return fut.result()
  File "uvloop/loop.pyx", line 1974, in create_connection
  File "uvloop/loop.pyx", line 1951, in uvloop.loop.Loop.create_connection
ConnectionRefusedError: [Errno 111] Connection refused

ERROR:    Application startup failed. Exiting.
(venv) quan@virt-centos:~/Vl-Agent$ echo $?                                                                                                    
0
```

</details>

### Environment

- OS / Python / Uvicorn version: just run `uvicorn --version`
- The exact command you're running uvicorn with, all flags you passed included. If you run it with gunicorn please do the same. If there is a reverse-proxy involved and you cannot reproduce without it please give the minimal config of it to reproduce.

**Comment by mboutet** [comment]
Hey @hongquan, I created a PR with a possible fix if you want to take a look at it.

**Comment by hongquan** [comment]
Resolved by #1278.
